test/functional/flow_controller_test.rb

```diff
@@ -34,4 +36,174 @@ class FlowControllerTest < ActionController::TestCase
+        get :update,
+            params: { id: "session-based",
+                      node_slug: "question1",
+                      response: "invalid" }
+
+        assert_redirected_to "/session-based/question1"
+      end
+
+      should "return an uncached response" do
+        get :update,
+            params: { id: "session-based",
+                      node_slug: "question1",
+                      response: "response1" }
+
+        assert_uncached_response
+      end
+    end
+
+    context "for a query parameter flow" do
+      should "redirect to the next question for a valid answer" do
+        get :update,
+            params: { id: "query-parameters-based",
+                      node_slug: "question1",
+                      response: "response1" }
+
+        assert_redirected_to "/query-parameters-based/question2?question1=response1"
+      end
+
+      should "redirect to the current question for an invalid answer" do
+        get :update,
+            params: { id: "query-parameters-based",
+                      node_slug: "question1",
+                      response: "invalid" }
+
+        assert_redirected_to "/query-parameters-based/question1?question1=invalid"
+      end
+
+      should "return a cached response" do
+        get :update,
+            params: { id: "query-parameters-based",
+                      node_slug: "question1",
+                      response: "response1" }
+
+        assert_cached_response
+      end
+    end
+
+    should "redirect to the first question with cache headers for a path based flow" do
+      get :update, params: { id: "radio-sample", node_slug: "anything" }
+      assert_redirected_to "/radio-sample/y"
+      assert_cached_response
+    end
+  end
+
+  context "GET /<id>/destroy_session" do
+    context "for a session flow" do
+      should "redirect to the first question without cache headers" do
+        get :destroy, params: { id: "session-based" }
+        assert_redirected_to "/session-based"
+        assert_uncached_response
+      end
+
+      should "remove responses from the session" do
+        session["session-based"] = { "question1" => "response1" }
+        get :destroy, params: { id: "session-based" }
+        assert_nil session["session-based"]
+      end
+    end
+
+    context "for a query parameter flow" do
+      should "redirect to the landing page with cache headers" do
+        get :destroy, params: { id: "query-parameters-based" }
+        assert_redirected_to "/query-parameters-based"
+        assert_cached_response
+      end
+
+      should "remove any answers from the query string" do
+        get :destroy, params: { id: "query-parameters-based", question1: "response1" }
+        assert_redirected_to "/query-parameters-based"
+      end
+    end
+
+    should "redirect to the first question with cache headers for a path based flow" do
+      get :destroy, params: { id: "radio-sample" }
+      assert_redirected_to "/radio-sample/y"
```

Comment:
It's strange that the other flows take you back to the start page, but the path-based flow goes back to the first question.

Reply:
Ooh.. you might not realise it but that just highlighted me to a test description issue: https://github.com/alphagov/smart-answers/compare/ae8906316f9aaea979a8c677d04ba14e1931f187..6f9fae6e2930aebf82bb22efc4584261ed93a62e thanks!

Yeah, this behaviour is different as the path based approach is part of a before_action filter (which, sidenote, I'm never sure of how best to test them - I'm a bit duplication heavy on it in this test) and it kind of makes sense to send them to the first question as they're most likely to experience this when a Smart Answer changes type.

Reply:
You're welcome?